app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusHandlerThread.java

```diff
@@ -330,66 +342,65 @@ public boolean processIncomingSleepData(byte[] data){
         return true;
     }
 
-    private void requestNextSleepData() {
-        mGetSleepTime = GregorianCalendar.getInstance();
-        mGetSleepTime.add(GregorianCalendar.SECOND, SLEEP_SYNC_RETRY_PERIOD);
-
-        TransactionBuilder builder = new TransactionBuilder("requestSleepStats");
-        builder.write(mHPlusSupport.ctrlCharacteristic, new byte[]{HPlusConstants.CMD_GET_SLEEP});
-        builder.queue(mHPlusSupport.getQueue());
-    }
-
-
+    /**
+     * Process a message containing real time information
+     *
+     * @param data the message from the device
+     * @return boolean indicating success or fail
+     */
     public boolean processRealtimeStats(byte[] data) {
         HPlusDataRecordRealtime record;
 
         try{
             record = new HPlusDataRecordRealtime(data);
         } catch(IllegalArgumentException e){
             LOG.debug((e.getMessage()));
-            return true;
+            return false;
         }
-
-        if(record.same(prevRealTimeRecord))
+        //Skip duplicated messages as the device seems to send the same record multiple times
+        //This can be used to detect the user is moving (not sleeping)
+        if(prevRealTimeRecord != null && record.same(prevRealTimeRecord))
             return true;
 
         prevRealTimeRecord = record;
 
         getDevice().setBatteryLevel(record.battery);
-        getDevice().sendDeviceUpdateIntent(getContext());
 
-        //Skip when measuring
-        if(record.heartRate == 255) {
+        //Skip when measuring heart rate
+        //Calories and Distance are updated and these values will be lost.
+        //Because a message with a valid Heart Rate will be provided, this loss very limited
+        if(record.heartRate == ActivityKind.TYPE_NOT_MEASURED) {
             getDevice().setFirmwareVersion2("---");
             getDevice().sendDeviceUpdateIntent(getContext());
-            return true;
+        }else {
+            getDevice().setFirmwareVersion2("" + record.heartRate);
```

Comment:
That's pretty much how it should be already if you use addDeviceInfo(). You can add all your desired fields like that. Do you want to show this information directly, instead of collapsed by default?